### PR TITLE
network: Enable the Kill Request filter to honor Route-level configur…

### PR DESCRIPTION
**Pull request by xiaobinxbw**
network: Enable the Kill Request filter to honor Route-level configuration. (#14929)

Signed-off-by: Tommy Wang <xiaobinwang@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/master/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/http/kill_request/kill_request_filter.cc
Comment (htuch) [inline source/extensions/filters/http/kill_request/kill_request_filter.cc:42]: I think we eventually want to match with the actual name used in the filter instantiation, but this hasn't been fixed Envoy-wide to my knowledge (https://github.com/envoyproxy/envoy/issues/12575).

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by xiaobinxbw** [comment]
Thanks Harvey!

In that case, could you please help me to click on the "Merge pull request"
button?

I am not authorised to merge :)

Thanks,
Tommy

**Comment by htuch** [comment]
@xiaobinxbw the coverage CI failure looks legit. Is there any way to bump test coverage?

**Comment by xiaobinxbw** [comment]
> i

**Comment by xiaobinxbw** [comment]
Hey Harvey,

I've added another use case which should cover the branches in my added code. I am not sure why it's still complaining about the coverage. Is there any way I can generate a coverage report? Sorry for the dumb question as it's my very first oss work :)

btw, i accidentally closed the pull request and reopened it. I will need your help with the approval again.

Thanks,
Tommy

File: test/extensions/filters/http/kill_request/kill_request_filter_test.cc
Comment (htuch) [inline test/extensions/filters/http/kill_request/kill_request_filter_test.cc:106]: These names for tests are kind of unreadably long. Can you replace them with a shorter name and then add a single line `//` comment above providing the intuition that is currently captured in the name? Here and elsewhere in these tests. Thanks.

**Comment by htuch** [comment]
@xiaobinxbw you can see the coverage report at https://storage.googleapis.com/envoy-pr/f7c1716/coverage/source/extensions/filters/http/kill_request/kill_request_filter.cc.gcov.html, which you get to by clicking through to the CI details and selecting the GCS link under "linux_x64 coverage" -> "Upload coverage report". You can see it's missing functional coverage of the behaviors in this PR.

To generate reports locally, follow https://github.com/envoyproxy/envoy/blob/main/bazel/README.md#coverage-builds.

File: test/extensions/filters/http/kill_request/kill_request_filter_test.cc
Comment (xiaobinxbw) [reply test/extensions/filters/http/kill_request/kill_request_filter_test.cc:106]: Done

**Comment by xiaobinxbw** [comment]
> @xiaobinxbw you can see the coverage report at https://storage.googleapis.com/envoy-pr/f7c1716/coverage/source/extensions/filters/http/kill_request/kill_request_filter.cc.gcov.html, which you get to by clicking through to the CI details and selecting the GCS link under "linux_x64 coverage" -> "Upload coverage report". You can see it's missing functional coverage of the behaviors in this PR.
> 
> To generate reports locally, follow https://github.com/envoyproxy/envoy/blob/main/bazel/README.md#coverage-builds.

Thanks a lot for the instructions, Harvey. 

However, the two test cases actually covers the red part in the report, and i am not sure why the coverage report doesn't think this way. My guess is that the decoder_filter_callbacks_.route_->route_entry_ call is mocked, and the actually killing part (raise(SIGABRT);) is hijacked?

**Comment by xiaobinxbw** [comment]
Hey Harvey,

Just a note that i've run the dependency check locally with the command "ci/run_envoy_docker.sh 'ci/do_ci.sh deps'", and it succeeded. I don't think the latest test comment change could break the dependencies either. I haven't found a way to re-trigger this presubmit check without committing a code change, though. 

Thanks,
Tommy

**Comment by htuch** [comment]
@xiaobinxbw are you sure those paths are executed? I.e. verified with log messages and the tests. There may be some coverage corner cases with death tests IIRC, but just want to check you have validated that those paths are taken in the tests you describe.

**Comment by xiaobinxbw** [comment]
Hey Harvey,

Yes, I've done some experiments with the tests, and I am sure that the red
paths in the coverage report are actually covered. It's a bit tricky to
tell as adding extra logs to the paths doesn't tell the truth either.

Here is what I did:
1. in source/extensions/filters/http/kill_request/kill_request_filter.cc,
comment out the "raise(SIGABRT)" line #56 and run the test. Now all the
tests that are supposed to crash envoy fail, which means the block from
line #54 to #57 in kill_request_filter.cc are covered.

2. in source/extensions/filters/http/kill_request/kill_request_filter.cc,
add "is_kill_request = false;" inside the "if (per_route_kill_settings !=
nullptr)" block at line #49. Now the new test
KillRequestEnabledFromRouteLevelConfiguration fails, which is as expected
as we are mocking a non-null per_route_kill_settings value and expect the
kill_request_ to be assigned a value from the per-route config. This means
the block from line #48 to #51 in kill_request_filter.cc are also covered.

Hope this makes sense :)

Thanks,
Tommy

**Review by htuch — COMMENTED** [review]
LGTM modulo comments and adding a coverage exception to https://github.com/envoyproxy/envoy/blob/main/test/per_file_coverage.sh#L5. I'm pretty sure that there was some issue around coverage and death tests from ages ago but I can't recall (maybe @dnoe remembers).

File: source/extensions/filters/http/kill_request/kill_request_filter.h
Comment (xiaobinxbw) [reply source/extensions/filters/http/kill_request/kill_request_filter.h:104]: Done

File: source/extensions/filters/http/kill_request/kill_request_filter.cc
Comment (xiaobinxbw) [reply source/extensions/filters/http/kill_request/kill_request_filter.cc]: Done

**Comment by htuch** [comment]
@xiaobinxbw looks good, can you get CI format to pass? Thanks. You will also need to add the coverage exception I mention in https://github.com/envoyproxy/envoy/pull/14944#pullrequestreview-589172115

**Comment by xiaobinxbw** [comment]
> @xiaobinxbw looks good, can you get CI format to pass? Thanks. You will also need to add the coverage exception I mention in [#14944 (review)](https://github.com/envoyproxy/envoy/pull/14944#pullrequestreview-589172115)

@htuch Done!

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
